udp-reassembly-offload.md

```diff
@@ -29,4 +29,80 @@ The NDIS interface for URO is used for communication between TCPIP and the NDIS
 
 > **TODO**
 
+ntddndis.h
+```
+#if (NDIS_SUPPORT_NDIS690)
+//
+// values used in UDP receive offload
+//
+#define NDIS_OFFLOAD_PARAMETERS_URO_DISABLED            1
+#define NDIS_OFFLOAD_PARAMETERS_URO_ENABLED             2
+#endif // (NDIS_SUPPORT_NDIS690)
+
+...
+
+#if (NDIS_SUPPORT_NDIS690)
+    struct
+    {
+        UCHAR               IPv4;
+        UCHAR               IPv6;
+    } UdpReceiveOffload;
+#endif // (NDIS_SUPPORT_NDIS690)
+} NDIS_OFFLOAD_PARAMETERS, *PNDIS_OFFLOAD_PARAMETERS;
+
+...
+
+#if (NDIS_SUPPORT_NDIS690)
+typedef struct _NDIS_UDP_RECV_OFFLOAD
+{
+    struct
+    {
+        BOOLEAN Enabled;
+    } IPv4;
+    struct
+    {
+        BOOLEAN Enabled;
+    } IPv6;
+} NDIS_UDP_RECV_OFFLOAD, *PNDIS_UDP_RECV_OFFLOAD;
+#endif
+...
+
+#if (NDIS_SUPPORT_NDIS690)
+  //
+  // UDP Receive Offload
+  //
+  NDIS_UDP_RECV_OFFLOAD        UdpRecvOffload;
+#endif
+
+} NDIS_OFFLOAD, *PNDIS_OFFLOAD;
+```
+
+nbluro.w
+```
+
+
+#if NDIS_SUPPORT_NDIS690
+
+//
+// Per-NetBufferList information for UdpRecvSegCoalesceOffloadInfo.
+//
+typedef struct _NDIS_UDP_RSC_OFFLOAD_NET_BUFFER_LIST_INFO
+{
+    union
+    {
+        struct
+        {
+            ULONG SegCount: 16;
```

Comment:
Why do you need the segment count?

Reply:
why not?

Reply:
I wonder why the existing NDIS_UDP_RSC_OFFLOAD_NET_BUFFER_LIST_INFO has a segment count field if SegSize itself isn't sufficient.

Reply:
SegSize alone is sufficient, but maybe SegCount saves repetitive division operations on the data path? It also provides a consistent field with RSC. Worth revisiting.

Reply:
If the UDP payloads are not all the same length, then a SegSize and SegCount are insufficient to convey the sizes, and a list of lengths is needed.

Reply:
Only the final segment is allowed to be < SegSize

Reply:
If arbitrary segments are allowed to be arbitrary lengths, then we cannot deliver the datagrams intact without O(N) lengths in the OOB.

Reply:
Right. Remember the trouble we had with USO with a partial tail?

Reply:
How can the OOB store dynamically-sized data? I thought it was all fixed-length data that had to fit within a pointer-size (64-bits).

Reply:
The NBLs/NBs will implicitly have the length(s), right? So shouldn't segment size be enough?

Reply:
I just went and looked at nbluro.w and something is already defined there. Can we actually, safely change it after it's been published?
```c
//
// Per-NetBufferList information for UdpRecvSegCoalesceOffloadInfo.
//
typedef struct _NDIS_UDP_RSC_OFFLOAD_NET_BUFFER_LIST_INFO
{
    union
    {
        struct
        {
            USHORT SegCount;
            USHORT SegSize;
        } Receive;

        PVOID Value;
    };
} NDIS_UDP_RSC_OFFLOAD_NET_BUFFER_LIST_INFO, *PNDIS_UDP_RSC_OFFLOAD_NET_BUFFER_LIST_INFO;
```

Reply:
Perhaps we just stick with what's there, as is.

Reply:
Theoretically we should be able to change it since we've not documented it nor allowed any external dependencies to be taken on it, but I'm fine with leaving it, too.

Reply:
If we can make the existing definition work a benefit is that it's already published. External folks could immediately start consuming it.

Reply:
@nibanks The existing behavior doesn't concatenate the UDP packets together, it's only the UDP payloads, and as a single NBL, so there's no individual lengths unless there's a way to stuff a variable-length list of lengths in the OOB.